satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+			return nil
+		}
+		return rpcstatus.Error(rpcstatus.Unknown, err.Error())
+	}
+
+	exitStatus, err := endpoint.overlaydb.GetExitStatus(ctx, nodeID)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitFinishedAt != nil {
+		// TODO revisit this. Should check if signature was sent
+		completed := &pb.SatelliteMessage{Message: &pb.SatelliteMessage_ExitCompleted{ExitCompleted: &pb.ExitCompleted{}}}
+		err = stream.Send(completed)
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitInitiatedAt == nil {
+		request := &overlay.ExitStatusRequest{NodeID: nodeID, ExitInitiatedAt: time.Now().UTC()}
+		_, err = endpoint.overlaydb.UpdateExitStatus(ctx, request)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = stream.Send(&pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}})
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitLoopCompletedAt == nil {
+		err = stream.Send(&pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}})
+		return Error.Wrap(err)
+	}
+
+	// TODO possibly switch out for custom queue
+	var pending sync.Map
+	pendingLength := func() int {
+		count := 0
+		pending.Range(func(key interface{}, value interface{}) bool {
+			count++
+			return true
+		})
+
+		return count
+	}
+	// this will be 1 until GetIncomplete* methods no longer return values
+	var morePiecesFlag int32 = 1
+
+	var group errgroup.Group
+	group.Go(func() error {
+		ticker := time.NewTicker(endpoint.interval)
+		defer ticker.Stop()
+
+		for range ticker.C {
+			if pendingLength() == 0 {
+				incomplete, err := endpoint.db.GetIncompleteNotFailed(ctx, nodeID, endpoint.config.EndpointBatchSize, 0)
+				if err != nil {
+					return Error.Wrap(err)
+				}
+
+				if len(incomplete) == 0 {
+					incomplete, err = endpoint.db.GetIncompleteFailed(ctx, nodeID, endpoint.config.EndpointMaxFailures+1, endpoint.config.EndpointBatchSize, 0)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+				}
+
+				if len(incomplete) == 0 {
+					endpoint.log.Debug("no more pieces to transfer for node.", zap.String("node ID", nodeID.String()))
+					atomic.StoreInt32(&morePiecesFlag, 0)
+					break
+				}
+
+				for _, inc := range incomplete {
+					pointer, err := endpoint.metainfo.Get(ctx, string(inc.Path))
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					remote := pointer.GetRemote()
+
+					found := false
+					for _, piece := range remote.GetRemotePieces() {
+						if piece.NodeId == nodeID && piece.PieceNum == inc.PieceNum {
+							found = true
+						}
+					}
+					if !found {
+						endpoint.log.Debug("piece no longer held by node.", zap.String("node ID", nodeID.String()), zap.ByteString("path", inc.Path), zap.Int32("piece num", inc.PieceNum))
+						continue
+					}
+
+					redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
```

Comment:
We should handle removing the piece from the pointer if `len(remote.GetRemotePieces()) > redundancy.OptimalThreshold()` before we request a new node to transfer from the overlay. No need to transfer if we can just delete the piece entirely.

Reply:
Updated